charmcraft/config.py

```diff
@@ -18,167 +18,195 @@
-        )
-    elif error.validator == "format":
-        field_ref = get_field_reference(error.absolute_path)
-        msg = "Bad charmcraft.yaml content; the {} {}: got {!r}.".format(
-            field_ref, error.cause, error.instance
-        )
-    else:
-        # safe fallback
-        msg = error.message
-
-    raise CommandError(msg)
-
-
-@format_checker.checks("url", raises=ValueError)
 def check_url(value):
     """Check that the URL has at least scheme and net location."""
     if isinstance(value, str):
         url = urlparse(value)
         if url.scheme and url.netloc:
-            return True
-    raise ValueError("must be a full URL (e.g. 'https://some.server.com')")
+            return value
+    raise ValueError(
+        "must be a fully qualified URL (such as 'https://some.server.com')"
+    )
 
 
-@format_checker.checks("relative_path", raises=ValueError)
 def check_relative_paths(value):
     """Check that the received paths are all valid relative ones."""
     if isinstance(value, str):
         # check if it's an absolute path using POSIX's '/' (not os.path.sep, as the charm's
         # config is independent of the platform where charmcraft is running)
         if value and value[0] != "/":
-            return True
-    raise ValueError("must be a valid relative URL")
+            return value
+    raise ValueError("must be a valid relative path")
 
 
-@attr.s(kw_only=True, frozen=True)
-class CharmhubConfig:
-    """Configuration for all Charmhub related options."""
+def format_pydantic_error_location(loc):
+    """Format location."""
+    loc_parts = []
+    for loc_part in loc:
+        if isinstance(loc_part, str):
+            loc_parts.append(loc_part)
+        elif isinstance(loc_part, int):
+            # Integer indicates an index. Go
+            # back and fix up previous part.
+            previous_part = loc_parts.pop()
+            previous_part += f"[{loc_part}]"
+            loc_parts.append(previous_part)
+        else:
+            raise RuntimeError(f"unhandled loc: {loc_part}")
 
-    api_url = attr.ib(default="https://api.charmhub.io")
-    storage_url = attr.ib(default="https://storage.snapcraftcontent.com")
+    loc = ".".join(loc_parts)
+
+    # Filter out internal __root__ detail.
+    loc = loc.replace(".__root__", "")
+    return loc
+
+
+def format_pydantic_error_msg(msg):
+    """Format pydantic's error message field."""
+    # Replace shorthand "str" with "string".
+    msg = msg.replace("str type expected", "string type expected")
+    return msg
 
-    @classmethod
-    def from_dict(cls, source):
-        """Build from a raw dict."""
-        return cls(**source)
 
+def format_pydantic_errors(errors):
+    """Format errors.
 
-class BasicPrime(tuple):
-    """Hold the list of files to include, specified under parts/bundle/prime configs.
+    Example 1: Single error.
 
-    This is a intermediate structure until we have the full Lifecycle in place.
+    Bad charmcraft.yaml content:
```

Comment:
do we want to use doctests?

Reply:
I did add one doctest to document the schema in the module because @facundobatista asked for the schema to be documented.  But I'll leave it up to @facundobatista if he wants more pervasive doctests (and a more generalized enablement across the codebase).

Reply:
I don't like doctests :/. -1 to start using them.

Reply:
Removed